File: ads/list.js
import { inspect } from "util"
import { error } from "./util.js"

export class LNode {
  value
  prev = null; next = null

  constructor(value) { this.value = value }
}

export class List {
  #head = null
  #length = 0

  get length() { return this.#length }

  static from (it) {
    const lst = new List()
    for (const el of it) { lst.push(el) }
    return lst
  }

  [Symbol.iterator]() { return this.values() }

  entries() {
    function* entries(nd) {
      while (nd) { yield nd; nd = nd.next }
    }
    return entries(this.#head)
  }

  values() {
    function* values(nd) {
      while (nd) { yield nd.value; nd = nd.next }
    }
    return values(this.#head)
  }

  [inspect.custom](path, opts) { return `List(${[...this].join(", ")})` }

  // O(1) pushes an element to a head
  push(value) {
    const nd = new LNode(value)
    nd.next = this.#head
    this.#head = nd
    ++this.#length
    return this
  }

  // O(1) pops an element from a head
  pop() {
    if (this.#length === 0) { error("pop from empty list") }
    const nd = this.#head
    this.#head = this.#head.next
    --this.#length
    return nd.value
  }

  // O(1) peeks an element from a head
  peek() {
    if (this.#length === 0) { error("peek from empty list") }
    return this.#head.value
  }

  // O(n) deletes a matching element or undefined
  delete(value, eq = (a, b) => a === b) {
    let nd = this.#head
    if (nd) {
      if (eq(value, nd.value)) {
        this.#head = nd.next
        --this.#length
        return nd.value
      }
      while (nd.next) {
        if (eq(value, nd.next.value)) {
          const value = nd.next.value
          nd.next = nd.next.next
          --this.#length
          return value
        }
        nd = nd.next
      }
    }
  }

  // O(n) returns a matching element or undefined
  get(value, eq = (a, b) => a === b) {
    for (const nd of this.entries()) {
      if (eq(value, nd.value)) { return nd }
    }
  }

  // O(n) reverses a list in place
  reverse() {
    let prev = null, curr = this.#head, next
    while(curr) {
      next = curr.next
      curr.next = prev; prev = curr
      curr = next
    }
    this.#head = prev
  }
}

export class DList {
  #head = null; #tail = null
  #length = 0

  get length() { return this.#length }

  static from(it) {
    const lst = new DList()
    for (const el of it) { lst.pushTail(el) }
    return lst
  }

  [Symbol.iterator]() { return this.values() }

  entries(reverse = false) {
    function* entries(nd) {
      while (nd) { yield nd; nd = reverse ? nd.prev : nd.next }
    }
    return entries(reverse ? this.#tail : this.#head)
  }

  values(reverse = false) {
    function* values(nd) {
      while (nd) { yield nd.value; nd = reverse ? nd.prev : nd.next }
    }
    return values(reverse ? this.#tail : this.#head)
  }

  [inspect.custom]() { return `DList(${[...this].join(", ")})` }

  // O(1) pushes an element to a head
  push(value) {
    const nd = new LNode(value)
    if (this.#length === 0) { this.#head = this.#tail = nd }
    else { nd.next = this.#head; this.#head = this.#head.prev = nd }
    ++this.#length
    return this
  }

  // O(1) pushes an element to a tail
  pushTail(value) {
    const nd = new LNode(value)
    if (this.#length === 0) { this.#head = this.#tail = nd }
    else { nd.prev = this.#tail; this.#tail = this.#tail.next = nd }
    ++this.#length
    return this
  }

  // O(1) pops an element from a head
  pop() {
    if (this.#length === 0) { error("pop from empty dlist") }
    const nd = this.#head
    --this.#length
    if (this.#length === 0) { this.#head = this.#tail = null }
    else { this.#head = this.#head.next; this.#head.prev = null }
    return nd.value
  }

  // O(1) pops an element from a tail
  popTail() {
    if (this.#length === 0) { error("popTail from empty dlist") }
    const nd = this.#tail
    --this.#length
    if (this.length === 0) { this.#head = this.#tail = null }
    else { this.#tail = this.#tail.prev; this.#tail.next = null }
    return nd.value
  }

  // O(1) peeks an element from a head
  peek() {
    if (this.#length === 0) { error("peek from empty dlist") }
    return this.#head.value
  }

  // O(1) peeks an element from a tail
  peekTail() {
    if (this.#length === 0) { error("peekTail from empty dlist") }
    return this.#tail.value
  }

  // O(1) inserts an element after a node
  insert(node, value) {
    const nd = new LNode(value), next = node.next
    node.next = nd; nd.prev = node
    if (next) { nd.next = next; next.prev = nd }
    else { this.#tail = nd }
    ++this.#length
  }

  // O(1) deletes a node
  delete(node) {
    --this.#length
    if (this.#length === 0) { this.#head = this.#tail = null }
    else if (node === this.#head) {
      this.#head = this.#head.next; this.#head.prev = null
    } else if (node === this.#tail) {
      this.#tail = this.#tail.prev; this.#tail.next = null
    } else { node.prev.next = node.next; node.next.prev = node.prev }
  }

  // O(n) returns a matching element or undefined
  get(value, eq = (a, b) => a === b) {
    for (const nd of this.entries()) {
      if (eq(value, nd.value)) { return nd }
    }
  }
}

File: go/ads/list.go
package ads

import "fmt"

type Node struct {
  value int
  next, prev *Node
}

type List struct {
  head *Node
  length int
}

func (l *List) Length() int {
  return l.length
}

func (l *List) FromSlice(slc []int) {
  for _, val := range slc {
    l.Push(val)
  }
}

func (l *List) Backward() func(yield func(i, val int) bool) {
  i, n := 0, l.head
  return func(yield func(i, val int) bool) {
    for n != nil && yield(i, n.value) {
      n = n.next
      i++
    }
  }
}

// O(1)
func (l *List) Push(val int) {
  nd := &Node{value: val}
  nd.next = l.head
  l.head = nd
  l.length++
}

// O(1)
func (l *List) Peek() (int, error) {
  if l.head == nil {
    return 0, fmt.Errorf("peek from empty list")
  }
  return l.head.value, nil
}

// O(1)
func (l *List) Pop() (int, error) {
  if l.head == nil {
    return 0, fmt.Errorf("pop from empty list")
  }
  nd := l.head
  l.head = l.head.next
  l.length--
  return nd.value, nil
}

func (l *List) Reverse() {
  var prev, next *Node
  nd := l.head
  for nd != nil {
    next = nd.next
    nd.next = prev
    prev = nd
    nd = next
  }
  l.head = prev
}

type DList struct {
  head, tail *Node
  length int
}

func (l *DList) Length() int {
  return l.length
}

func (l *DList) FromSlice(slc []int) {
  for _, val := range slc {
    l.PushHead(val)
  }
}

func (l *DList) Backward() func(yield func(i, val int) bool) {
  i, nd := 0, l.head
  return func(yield func(i, val int) bool) {
    for nd != nil && yield(i, nd.value) {
      nd = nd.next
      i++
    }
  }
}

func (l *DList) Forward() func(yield func(i, val int) bool) {
  i, nd := 0, l.tail
  return func(yield func(i, val int) bool) {
    for nd != nil && yield(i, nd.value) {
      nd = nd.prev
      i++
    }
  }
}

// O(1)
func (l *DList) PushHead(val int) {
  nd := &Node{value: val}
  l.length++
  if l.head == nil {
    l.head, l.tail = nd, nd
    return
  }
  nd.next = l.head
  l.head.prev = nd
  l.head = nd
}

// O(1)
func (l *DList) PushTail(val int) {
  nd := &Node{value: val}
  l.length++
  if l.tail == nil {
    l.head, l.tail = nd, nd
    return
  }
  nd.prev = l.tail
  l.tail.next = nd
  l.tail = nd
}

// O(1)
func (l *DList) PeekHead() (int, error) {
  if l.head == nil {
    return 0, fmt.Errorf("peek head from empty dlist")
  }
  return l.head.value, nil
}

// O(1)
func (l *DList) PeekTail() (int, error) {
  if l.tail == nil {
    return 0, fmt.Errorf("peek tail from empty dlist")
  }
  return l.tail.value, nil
}

// O(1)
func (l *DList) PopHead() (int, error) {
  if l.head == nil {
    return 0, fmt.Errorf("pop head from empty dlist")
  }
  nd := l.head
  l.length--
  if l.head.next == nil {
    l.head, l.tail = nil, nil
    return nd.value, nil
  }
  l.head = l.head.next
  l.head.prev = nil
  return nd.value, nil
}

// O(1)
func (l *DList) PopTail() (int, error) {
  if l.tail == nil {
    return 0, fmt.Errorf("pop tail from empty dlist")
  }
  nd := l.tail
  l.length--
  if l.tail.prev == nil {
    l.head, l.tail = nil, nil
    return nd.value, nil
  }
  l.tail = l.tail.prev
  l.tail.next = nil
  return nd.value, nil
}

// O(1)
func (l *DList) Insert(val int, nd *Node) {
  newnd := &Node{value: val}
  l.length++
  newnd.next = nd.next
  newnd.prev = nd
  if nd == l.tail {
    l.tail = newnd
  } else {
    nd.next.prev = newnd
  }
  nd.next = newnd
}

// O(1)
func (l *DList) Delete(nd *Node) {
  l.length--
  if nd == l.head {
    l.head = l.head.next
    l.head.prev = nil
    return
  }
  if nd == l.tail {
    l.tail = l.tail.prev
    l.tail.next = nil
    return
  }
  nd.prev.next = nd.next
  nd.next.prev = nd.prev
}
